Tests/Unit/NetworkServer/CupsProtocolMessageProcessorTests.cs

```diff
@@ -110,21 +118,167 @@ public async Task HandleUpdateInfoAsync_Fails_WithInvalidInput(string input, Typ
             Assert.Contains(this.logger.Invocations, i => i.Arguments.Any(a => a.GetType() == exceptionType));
         }
 
+        [Fact]
+        public async Task HandleUpdateInfoAsync_Invokes_FetchFirmware_WhenUpdateAvailable()
+        {
+            // setup
+            using var memoryStream = new MemoryStream();
+            var strictifiedInput = JsonUtil.Strictify(CupsRequestJson);
+            var (httpContext, httpRequest, _) = SetupHttpContextWithRequest(strictifiedInput, memoryStream);
+            _ = httpRequest.Setup(r => r.ContentLength).Returns(Encoding.UTF8.GetByteCount(strictifiedInput));
+
+            // setting up the twin in such a way that there is a fw update but no matching checksum
+            var cupsTwinInfo = new CupsTwinInfo(new Uri(CupsUri),
+                                                new Uri(TcUri),
+                                                CredentialsChecksum,
+                                                CredentialsChecksum,
+                                                string.Empty,
+                                                string.Empty,
+                                                "anotherVersion",
+                                                KeyChecksum,
+                                                "ABCD");
+            _ = this.basicsStationConfigurationService.Setup(m => m.GetCupsConfigAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                                      .Returns(Task.FromResult(cupsTwinInfo));
+
+            var firmwareBytes = new byte[] { 1, 2, 3 };
+            using var fwStream = new MemoryStream(firmwareBytes);
+            using var streamContent = new StreamContent(fwStream);
+            _ = this.deviceAPIServiceBase.Setup(m => m.FetchStationFirmwareAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
+                                         .ReturnsAsync(streamContent);
```

Comment:
You can use a `ByteArrayContent` and skip the memory stream entirely (ditto for other tests):

```suggestion
            using var httpContent = new ByteArrayContent(firmwareBytes);
            _ = this.deviceAPIServiceBase.Setup(m => m.FetchStationFirmwareAsync(It.IsAny<StationEui>(), It.IsAny<CancellationToken>()))
                                         .ReturnsAsync(httpContent);
```